attribute/yoti.go

```diff
@@ -0,0 +1,79 @@
+package attribute
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/getyoti/yoti-go-sdk/anchor"
+)
+
+// Attribute represents a small piece of information about a Yoti user such as a photo of the user or the
+// user's date of birth.
+type Attribute struct {
+	Name    string
+	Value   []byte
+	Type    AttrType
+	Anchors []*anchor.Anchor
+}
+
+//AttrType format of the attribute
+type AttrType uint
+
+const (
+	//AttrTypeTime date format
+	AttrTypeTime AttrType = iota + 1
+	//AttrTypeString text format
+	AttrTypeString
+	//AttrTypeJPEG JPEG format
+	AttrTypeJPEG
+	//AttrTypePNG PNG fornmat
+	AttrTypePNG
+	//AttrTypeJSON JSON fornmat
+	AttrTypeJSON
+	//AttrTypeBool Boolean format
+	AttrTypeBool
+	//AttrTypeInterface generic interface format
+	AttrTypeInterface
+)
+
+// AttrValue represents the value associated with a Yoti Attribute.
+type AttrValue struct {
+	// Type represents the format of the piece of user data, whether it is a date, a piece of text or a picture
+	//
+	// Note the potential values for this variable are stored in constants with names beginning with
+	// 'AttrType'. These include:
+	//  yoti.AttrTypeTime
+	//  yoti.AttrTypeString
+	//  yoti.AttrTypeJPEG
+	//  yoti.AttrTypePNG
+	//  yoti.AttrTypeJSON
+	//  yoti.AttrTypeBool
+	//  yoti.AttrTypeInterface
+	Type  AttrType
+	Value []byte
+}
+
+// GetMIMEType returns the MIME type of this piece of Yoti user information. For more information see:
+// https://en.wikipedia.org/wiki/Media_type
+func GetMIMEType(attributeType fmt.Stringer) (result string) {
+	switch attributeType {
+	case AttrTypeTime:
+		result = "text/plain; charset=UTF-8"
+
+	case AttrTypeString:
+		result = "text/plain; charset=UTF-8"
+
+	case AttrTypeJPEG:
+		result = "image/jpeg"
+
+	case AttrTypePNG:
+		result = "image/png"
+
+	case AttrTypeJSON:
+		result = "application/json; charset=UTF-8"
+
+	default:
+		log.Printf("Unable to find a matching MIME type for value type %q", attributeType.String())
+	}
+	return
+}
```

Comment:
What are you gaining with this file? I really dont think this gives you much more than code to maintain. This GetMIMEType function is probably the most useful part, maybe put it in a map?

Reply:
I think I have removed this intermediary type at one point of this work, and it made testing a lot more difficult - having to create these protobuf attributes instead of a new type. 

Additionally if we had the protobuf attribute, we'd have the raw anchors on this. Since we don't want to expose these to the clients, this would mean we'd have to parse these anchors to extract the information we need at the point of the user accessing the attribute from the profile. I believe we want to be doing this before this stage.

Reply:
Where would we make the map for MimeType if we were to make one? I think an extension method is quite appropriate for its use

Reply:
"it made testing a lot more difficult - having to create these protobuf attributes instead of a new type." - I don't see how this is different, you will still need to construct the types for testing... also you could build test helpers. Core has some Go code for doing this, maybe we could share?

Reply:
can make the map right here: 
```
func GetMIMEType(attributeType AttrType) string {
    return map[AttrType]string{
                {AttrTypeTime: "text/plain; charset=UTF-8"},
               ...
    }

Reply:
> "it made testing a lot more difficult - having to create these protobuf attributes instead of a new type." - I don't see how this is different, you will still need to construct the types for testing... also you could build test helpers. Core has some Go code for doing this, maybe we could share?
> With regards to the second part, I'm not sure why including the 'raw' anchor is an issue, the only interesting information in there is what we are going to share and the YOTI signatures to verify that YOTI has processed the data.... this is all OK for the public to see.

If you have some Go code for building test helpers, that would be much appreciated. If you could reference them in [the ticket for looking at using Protobuf types]( https://lampkicking.atlassian.net/browse/SDK-624), then we can get this PR merged, and look at doing this as part of that ticket.

Reply:
> can make the map right here:
> 
> ```
> func GetMIMEType(attributeType AttrType) string {
>     return map[AttrType]string{
>                 {AttrTypeTime: "text/plain; charset=UTF-8"},
>                ...
>     }
> Also notice the more specific type being passed in
> ```

I did have the more specific type originally, but gometalinter suggested I change it  to `fmt.Stringer`(not sure why). I'll change it back

Reply:
![image](https://user-images.githubusercontent.com/1381991/47493790-0b157280-d848-11e8-983f-c1446a197aef.png)

Reply:
oh, ok. That's interesting. Not sure why its making that suggestion. Going to ask @simon-yoti

Reply:
Is there any more to this warning message?

Reply:
No, that is it unfortunately

Reply:
I think this is because of the Go mantra
>return types, accept interfaces.

So the linter looks to see if an argument implements an interface, and if it does suggests you use that instead. In this case it's wrong; `AttrType` does have a `String() string` method and so implements `fmt.Stringer`, but the behaviour of the function has nothing to with what interface it implements (it just happens to use the `String` method in an error message). Linters aren't always right, that's why they're linters.

Reply:
Thanks, ok I'll leave it as using `AttrType`, not `fmt.Stringer`